public/app/features/dashboard/components/PanelEditor/utils.ts

```diff
@@ -48,19 +48,40 @@ export const updateDefaultFieldConfigValue = (
       if (remove) {
         defaults.custom = omit(defaults.custom, name);
       } else {
-        defaults.custom = lodashSet({ ...defaults.custom }, name, value);
+        defaults.custom = setOptionImmutably({ ...defaults.custom }, name, value);
       }
     } else if (!remove) {
-      defaults.custom = lodashSet({ ...defaults.custom }, name, value);
+      defaults.custom = setOptionImmutably({ ...defaults.custom }, name, value);
     }
   } else if (remove) {
     defaults = omit(defaults, name);
   } else {
-    defaults = lodashSet({ ...defaults }, name, value);
+    defaults = setOptionImmutably({ ...defaults }, name, value);
   }
 
   return {
     ...config,
     defaults,
   };
 };
+
+export function setOptionImmutably<T extends object>(options: T, path: string, value: any): T {
+  const splat = path.split('.');
+  let result = { ...options };
+
+  if (splat.length === 1) {
+    return {
+      ...result,
+      [path]: value,
+    };
+  }
+
+  let current = (result as Record<string, any>)[splat[0]];
+  if (current === undefined || current === null) {
+    current = {} as object;
+  }
+  return {
+    ...result,
+    [splat[0]]: setOptionImmutably({ ...current }, splat.slice(1).join('.'), value),
+  };
+}
```

Comment:
I am afraid of the performance with this, wonder if there is a way to optimize this.

Reply:
Friendly pinging @leeoniya to chime in ;)

Reply:
US holiday today, but will take a look tonight.

Reply:
i don't think this will have any impact on perf since:

1. it looks like it's called only during panel option edits (i checked by wrapping lodashSet in a function with `console.count()` and playing with a few gdev dashboards)
2. 99% of the time the path length is 1, so we're only copying 1 level, as we did before. there'll some microseconds lost on doing the extra `splat.shift()` i added, but it won't be relevant.

Reply:
ah, that is true. I  thought this was called for every field options during applyFieldOverrides (hence my concern)